rfcs/0010-mempool-dos-mitigation.md

```diff
@@ -0,0 +1,302 @@
+## Mitigations for DOS attacks based on bogus transactions
+[summary]: #summary
+
+We add some rules for deciding whether to store incoming transactions in the
+mempool and gossip them. FIXME details.
+
+
+## Motivation
+
+[motivation]: #motivation
+
+Every transaction consumes resources: computation, storage, and bandwidth. In
+the case of transactions that are eventually included in a block, those
+resources are priced by transaction fees. But transactions that are never
+included in a block don't pay transaction fees, so without good design bogus
+transactions could be a vector for cheap DoS attacks. We want to make sure it
+costs enough to make nodes consume resources that a DoS attack is too expensive
+to be worthwhile.
+
+(The transaction fees for *mined* transactions may or may not accurately reflect
+the total cost to all network participants for processing the transaction, but
+that's a problem for another day.)
+
+We can't control what transactions we receive, but we can control whether we
+store them and whether we forward them on, making other nodes deal with them.
+
+## Detailed design
+
+[detailed-design]: #detailed-design
+
+As a general principle, we want to accept transactions into the mempool iff they
+will eventually be mined into a future block. That's the purpose of the mempool.
+For DoS protection, we assume that the cost of including a transaction in a
+block is sufficient to deter DoS attacks on proposers, since the cost of
+creating a SNARK greatly exceeds the cost of checking and storing an incoming
+transaction. Where possible, we prefer to charge for things rather than banning
+them. We don't mind if someone uses a lot of resources, so long as those
+resources are paid for at a price that network participants would be happy with.
+
+### When we first recieve a transaction: rules
+
+When we receive a gossiped transaction, we will check the below constraints. If
+any of the checks fail, we ignore the transaction and do not gossip it.
+
+* Validity:
```

Comment:
How many of these checks and sanctions are already implemented?

Reply:
We only check signatures.